direct-messages/README.md

```diff
@@ -10,24 +10,74 @@ because the same message can simultaneously:
+- `scalarmult`: for example, the curve25519 scalar multiplication function. As common in elliptic curve cryptography, we use additive group notation. Should the underlying Diffie-Hellman algorithms be exchanged, the call is substituted with the respective scalar multiplication or exponentiation function.
+- `hkdf`: The HKDF function, as specified in [RFC 5869]
+- `tfk_encode`: The binary encoding of keys, as specified in TODO
+
+As a high level overview, we first compute the shared key between our secret key and the other public key.
+Then, we derive a secret from it using HKDF.
+The key derivation function takes four inputs: the input keying material, the key length, a salt and some additional info that ties the output key to the context.
+A different info value will make the key derivation return a different value, even if the same input key is used.
+This means the `info` variable can be used to tie the output key to the specific context of this call. 
+That context is composed of several values.
+
+Firstly, it contains a purpose identifying this particular key derivation.
+
+Secondly, it contains the two Diffie-Hellman public keys whose shared secret is calculated, concatenated with the respective feed ID.
+  We include the Diffie-Hellman public keys, because it is possible for two pairs `(skA, pkB), (skC, pkD)` to have scalar multiplication result in the same shared key `skA*pkB=skC*pkD`.
+  By deriving based on the two public keys, we disambiguate which of the possible keys were used and return a different output key, respectively.
+  Note that this can only happen if the attacker has access to the secret keys, but even under these circumstances we want to avoid the confusion that can arise from this.
+  The feed IDs are included to avoid any confusion that could arise by an attacker reposting someone elses public keys.
+  Concatenating them like this makes sure that the Diffie-Hellman public key and feed ID remain coupled.
+  If we would concatenate them separately, it would not be immediately clear which Diffie-Helman public key belongs to which feed ID.
+  Also, these two values are sorted.
+  This means that the order in which they are supplied is not relevant for the output encoding, and both parties will always result at the same encoding and output key if they submit the same keys.
+
+This way we
+- specify the purpose of the key derivation
+- specify the public keys whose shared secret is the input keying material
+- specify the feed IDs of the parties for who the DM keys should be computed
+- public keys and feed IDs are coupled, such that they can not be swapped
+
+### Note:
+We (curently) use the feed (signing) keys for this derivation. since these are `ed25519` keypairs that can't do Diffie-Hellman, we convert these to `curve25519` keypairs before doing scalarmult. In the medium term we intend to introduce messages that allow to override the own Diffie-Hellman public key used by others.
```

Comment:
Therefore it is important that we are able to specify _which_ public key we are using in association with which feedId (e.g. `@ascqFASasda4asdasd=.ed25519`). 

Arg ... I think perhaps a section somewhere on how `feedId`'s are primarily about signing... and we just happen to also use them to generate a DH-public key for private messages. I'm realising it's too late for me to be writing good copy (text)